api/simulate.go

```diff
@@ -0,0 +1,324 @@
+		parent = result.Header()
+
+		s.state.Finalise()
+	}
+
+	return results, nil
+}
+
+func (s *simulator) processBlock(ctx *rpc.CallContext, block *simBlock, header, parent *types.Header, headers []*types.Header, timeout time.Duration) (*types.Block, []simCallResult, map[ctypes.Hash]common.Address, error) {
+	// Set header fields that depend only on parent block.
+	// Parent hash is needed for evm.GetHashFn to work.
+	header.ParentHash = parent.Hash()
+	if s.chainConfig.IsLondon(header.Number) {
+		if header.BaseFee == nil {
+			if s.validate {
+				header.BaseFee = eip1559.CalcBaseFee(s.chainConfig, parent)
+			} else {
+				header.BaseFee = big.NewInt(0)
+			}
+		}
+	}
+	if s.chainConfig.IsCancun(header.Number, new(big.Int).SetUint64(header.Time)) {
+		var excess uint64
+		if s.chainConfig.IsCancun(parent.Number, new(big.Int).SetUint64(parent.Time)) {
+			parentExcess := uint64(0)
+			if parent.ExcessBlobGas != nil {
+				parentExcess = *parent.ExcessBlobGas
+			}
+			parentBlobGasUsed := uint64(0)
+			if parent.BlobGasUsed != nil {
+				parentBlobGasUsed = *parent.BlobGasUsed
+			}
+			excess = eip4844.CalcExcessBlobGas(parentExcess, parentBlobGasUsed)
+		}
+		header.ExcessBlobGas = &excess
+	}
+	
+
+	// State overrides are applied prior to execution of a block
+	if block.StateOverrides != nil {
+		if err := block.StateOverrides.Apply(s.state); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
+	var (
+		blobGasUsed uint64
+		gasUsed     uint64
+		txes        = make([]*types.Transaction, len(block.Calls))
+		callResults = make([]simCallResult, len(block.Calls))
+		receipts    = make([]*types.Receipt, len(block.Calls))
+		senders     = make(map[ctypes.Hash]common.Address)
+		// allLogs     []*types.Log
+	)
+
+	getHashFn := func(n uint64) ctypes.Hash {
+		for _, h := range headers {
+			if h.Number.Uint64() == n {
+				return h.Hash()
+			}
+		}
+		if parent.Number.Uint64() == n {
+			return parent.Hash()
+		}
+		return ctypes.Hash{}
+	}
+
+	for i, call := range block.Calls {
+		if err := ctx.Context().Err(); err != nil {
+			return nil, nil, nil, err
+		}
+		if err := call.setDefaults(ctx, s.evmFn, s.state, header, vm.BlockNumberOrHashWithHash(header.Hash(), false)); err != nil {
+			return nil, nil, nil, err
+		}
+		if gasUsed+uint64(*call.Gas) > header.GasLimit {
+			return nil, nil, nil, fmt.Errorf("block gas limit exceeded")
+		}
+		tx := call.ToTransaction(types.DynamicFeeTxType)
+		txes[i] = tx
+		senders[tx.Hash()] = call.from()
+		s.state.SetTxContext(tx.Hash(), i)
+
+		evm := s.evmFn(s.state, &vm.Config{
+			NoBaseFee: !s.validate,
+		}, call.from(), call.GasPrice.ToInt())
+
+		if evm.Context.GetHash == nil {
+			evm.Context.GetHash = getHashFn
+		}
+
+		if s.chainConfig.IsPrague(header.Number, new(big.Int).SetUint64(header.Time)) { 
+			// Process parent block hash for EIP-2935
+			if header.ParentHash != (ctypes.Hash{}) {
+				evm.StateDB.SetState(params.HistoryStorageAddress,ctypes.BigToHash(header.Number), header.ParentHash,)
```

Comment:
I believe this should be unnecessary, because the state updates should be streamed to Cardinal from the masters.